apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/pac/FissClaimTransformerV2.java

```diff
@@ -268,37 +269,89 @@ private Reference getFacility(RdaFissClaim claimGroup) {
   private List<Claim.DiagnosisComponent> getDiagnosis(RdaFissClaim claimGroup, boolean isIcd9) {
     final String icdSystem = isIcd9 ? IcdCode.CODING_SYSTEM_ICD_9 : IcdCode.CODING_SYSTEM_ICD_10_CM;
 
-    return ObjectUtils.defaultIfNull(claimGroup.getDiagCodes(), List.<RdaFissDiagnosisCode>of())
-        .stream()
-        .map(
-            diagnosisCode -> {
-              Claim.DiagnosisComponent component;
+    Stream<FissDiagnosisAdapterV2> diagnosisCodes =
+        ObjectUtils.defaultIfNull(claimGroup.getDiagCodes(), List.<RdaFissDiagnosisCode>of())
+            .stream()
+            .sorted(Comparator.comparing(RdaFissDiagnosisCode::getRdaPosition))
+            .map(FissDiagnosisAdapterV2::new);
+
+    Stream<FissDiagnosisAdapterV2> combinedDiagnosisCodes =
+        getAllDistinctDiagnosisCodes(claimGroup, diagnosisCodes);
+
+    List<Claim.DiagnosisComponent> diagnoses =
+        combinedDiagnosisCodes
+            .map(diagnosisCode -> getDiagnosisComponent(diagnosisCode, claimGroup, icdSystem))
+            .filter(Objects::nonNull)
+            .toList();
+
+    // Certain records may be filtered out, so we need to set the sequence after we finish
+    // processing.
+    for (int i = 0; i < diagnoses.size(); i++) {
+      diagnoses.get(i).setSequence(i + 1);
+    }
+    return diagnoses;
+  }
 
-              if (Strings.isNotBlank(diagnosisCode.getDiagCd2())) {
-                component =
-                    new Claim.DiagnosisComponent()
-                        .setSequence(diagnosisCode.getRdaPosition())
-                        .setDiagnosis(createCodeableConcept(icdSystem, diagnosisCode.getDiagCd2()));
-
-                if (Strings.isNotBlank(diagnosisCode.getDiagPoaInd())) { // Present on Admission
-                  component.setOnAdmission(
-                      createCodeableConcept(
-                          BBCodingSystems.FISS.DIAG_POA_IND,
-                          diagnosisCode.getDiagPoaInd().toLowerCase()));
-                }
+  /**
+   * Creates a distinct list of all diagnosis codes for the claim.
+   *
+   * @param claimGroup FISS claim.
+   * @param diagnosisCodes List of codes.
+   * @return All diagnosis codes wrapped in adapters.
+   */
+  private static Stream<FissDiagnosisAdapterV2> getAllDistinctDiagnosisCodes(
+      RdaFissClaim claimGroup, Stream<FissDiagnosisAdapterV2> diagnosisCodes) {
+    Stream<FissDiagnosisAdapterV2> extraCodes =
+        Stream.of(
+            new FissDiagnosisAdapterV2(claimGroup.getAdmitDiagCode()),
+            new FissDiagnosisAdapterV2(claimGroup.getPrincipleDiag()));
+
+    // Most of the time, the principal and admit codes should already be included in the list
+    // returned by getDiagCodes(),
+    // but if they're not, we need to make sure they're included in the final result.
+    // Here we'll concatenate them all together and filter out any duplicates.
+    return Stream.concat(diagnosisCodes, extraCodes).distinct();
```

Comment:
One place where this _could_ be different than the current behavior is if there are duplicate codes in the main diagnosis code list. Currently those would just be returned as is, but here the duplicates would be removed. I don't see why anyone would actually want to see the same code returned twice though so I don't think this is an issue.